src/dma/mod.rs

```diff
@@ -465,133 +441,159 @@ where
             }
-
-            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
-                unsafe {
-                    self.stream.set_memory_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_address() != new_buf_ptr as u32 {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+            Some(inactive) => {
+                // Double buffer mode
+                if !STREAM::get_transfer_complete_flag() {
+                    // DMA has not released a buffer
+                    return Err(DMAError::NotReady);
                 }
-
-                // "Subsequent reads and writes cannot be moved ahead of
-                // preceding reads"
-                compiler_fence(Ordering::Acquire);
-
-                let old_buf = self.buf.replace(new_buf);
-
-                // We always have a buffer, so unwrap can't fail
-                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
-            } else {
+                // Poison the peripheral's inactive memory address to get a memory
+                // error instead of potentially silent corruption.
+                // If DMA wins the race (overrun) to the inactive buffer
+                // between reading the CT bit and poisoning the inactive address, this
+                // write will fail and lead to a transfer error (TEIF) and disable
+                // the stream.
+                // If DMA wins the race by the time we write the new valid address
+                // (below), it gets a bus error and errors/stops.
                 unsafe {
-                    self.stream
-                        .set_memory_double_buffer_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_double_buffer_address()
-                    != new_buf_ptr as u32
-                {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+                    self.stream.set_memory_address(inactive, 0xffff_ffffusize);
                 }
+                (false, inactive)
+            }
+        };
 
-                // "Subsequent reads and writes cannot be moved ahead of
-                // preceding reads"
-                compiler_fence(Ordering::Acquire);
+        // Protect the instruction sequence of preceding DMA disable/inactivity
+        // verification/poisoning and subsequent (old completed) buffer content
+        // access.
+        // Cortex-M7: Also protect the corresponding data access sequence.
+        // NOTE: The data cache also needs to be flushed (if enabled).
```

Comment:
It seems to me like this is something we should be doing for all users, since the DMA memory is guaranteed to be in a cache-able memory location. We should do an MVA invalidate for all addresses in the buffer if the d-cache is enabled, as the comment here indicates.

Note that there seems to be some issue here because `cortex-m` requires ownership of the SCB for `invalidate_dcache_by_address()` and cache lines must be 32-byte aligned, which means that DMA buffers must be cache-aligned when the d-cache is enabled as well.

Now that I've written all of this, it may be worth breaking out into a separate PR to add support for d-cache-enabled implementations.

https://docs.rs/cortex-m/0.7.0/cortex_m/peripheral/struct.SCB.html#method.invalidate_dcache_by_address

Reply:
It looks like that would also restrict us to using buffers that are a length that is a multiple of the cache size.

Reply:
Note that there's a lot of other cache operations that we aren't supporting (such as cleaning the active buffer to commit it to memory before starting a DMA write), so this isn't necessarily the only place where cache operation support needs to be added. Let's leave this for the future though and add a disclaimer that DMA does not support the D-cache being enabled.